File: electron/database/wikiTagRepository.ts
import type Database from "better-sqlite3";
import type { WikiTag, WikiTagAssignment } from "../types";

interface WikiTagRow {
  id: string;
  name: string;
  color: string;
  text_color: string | null;
  created_at: string;
  updated_at: string;
}

interface WikiTagAssignmentRow {
  tag_id: string;
  entity_id: string;
  entity_type: string;
  source: string;
  created_at: string;
}

function rowToTag(row: WikiTagRow): WikiTag {
  return {
    id: row.id,
    name: row.name,
    color: row.color,
    textColor: row.text_color ?? undefined,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function rowToAssignment(row: WikiTagAssignmentRow): WikiTagAssignment {
  return {
    tagId: row.tag_id,
    entityId: row.entity_id,
    entityType: row.entity_type as WikiTagAssignment["entityType"],
    source: row.source as WikiTagAssignment["source"],
    createdAt: row.created_at,
  };
}

export function createWikiTagRepository(db: Database.Database) {
  const stmts = {
    fetchAll: db.prepare(`SELECT * FROM wiki_tags ORDER BY name ASC`),
    fetchById: db.prepare(`SELECT * FROM wiki_tags WHERE id = ?`),
    fetchByName: db.prepare(`SELECT * FROM wiki_tags WHERE name = ?`),
    search: db.prepare(
      `SELECT * FROM wiki_tags WHERE name LIKE '%' || ? || '%' ORDER BY name ASC`,
    ),
    insert: db.prepare(`
      INSERT INTO wiki_tags (id, name, color, created_at, updated_at)
      VALUES (@id, @name, @color, @created_at, @updated_at)
    `),
    insertOrReplace: db.prepare(`
      INSERT OR REPLACE INTO wiki_tags (id, name, color, created_at, updated_at)
      VALUES (@id, @name, @color, @created_at, @updated_at)
    `),
    update: db.prepare(`
      UPDATE wiki_tags SET name = @name, color = @color, text_color = @text_color, updated_at = @updated_at
      WHERE id = @id
    `),
    delete: db.prepare(`DELETE FROM wiki_tags WHERE id = ?`),
    // Assignment statements
    fetchTagsForEntity: db.prepare(`
      SELECT wt.* FROM wiki_tags wt
      INNER JOIN wiki_tag_assignments wta ON wt.id = wta.tag_id
      WHERE wta.entity_id = ?
      ORDER BY wt.name ASC
    `),
    fetchAssignmentsForEntity: db.prepare(`
      SELECT * FROM wiki_tag_assignments WHERE entity_id = ?
    `),
    deleteAssignmentsForEntity: db.prepare(
      `DELETE FROM wiki_tag_assignments WHERE entity_id = ?`,
    ),
    deleteInlineAssignmentsForEntity: db.prepare(
      `DELETE FROM wiki_tag_assignments WHERE entity_id = ? AND source = 'inline'`,
    ),
    insertAssignment: db.prepare(`
      INSERT OR IGNORE INTO wiki_tag_assignments (tag_id, entity_id, entity_type, source, created_at)
      VALUES (@tag_id, @entity_id, @entity_type, @source, @created_at)
    `),
    fetchAllAssignments: db.prepare(`SELECT * FROM wiki_tag_assignments`),
    // For merge
    updateAssignmentTagId: db.prepare(`
      UPDATE OR IGNORE wiki_tag_assignments SET tag_id = @new_tag_id WHERE tag_id = @old_tag_id
    `),
    deleteAssignmentsByTag: db.prepare(
      `DELETE FROM wiki_tag_assignments WHERE tag_id = ?`,
    ),
  };

  return {
    fetchAll(): WikiTag[] {
      return (stmts.fetchAll.all() as WikiTagRow[]).map(rowToTag);
    },

    search(query: string): WikiTag[] {
      return (stmts.search.all(query) as WikiTagRow[]).map(rowToTag);
    },

    createWithId(id: string, name: string, color: string): WikiTag {
      const now = new Date().toISOString();
      stmts.insertOrReplace.run({
        id,
        name,
        color,
        created_at: now,
        updated_at: now,
      });
      const row = stmts.fetchById.get(id) as WikiTagRow;
      return rowToTag(row);
    },

    create(name: string, color: string): WikiTag {
      const now = new Date().toISOString();
      const id = `tag-${crypto.randomUUID()}`;
      stmts.insert.run({
        id,
        name,
        color,
        created_at: now,
        updated_at: now,
      });
      const row = stmts.fetchById.get(id) as WikiTagRow;
      return rowToTag(row);
    },

    update(
      id: string,
      updates: Partial<Pick<WikiTag, "name" | "color" | "textColor">>,
    ): WikiTag {
      const existing = stmts.fetchById.get(id) as WikiTagRow | undefined;
      if (!existing) throw new Error(`WikiTag not found: ${id}`);
      const now = new Date().toISOString();
      stmts.update.run({
        id,
        name: updates.name ?? existing.name,
        color: updates.color ?? existing.color,
        text_color:
          "textColor" in updates
            ? (updates.textColor ?? null)
            : existing.text_color,
        updated_at: now,
      });
      const row = stmts.fetchById.get(id) as WikiTagRow;
      return rowToTag(row);
    },

    delete(id: string): void {
      stmts.delete.run(id);
    },

    merge(sourceId: string, targetId: string): WikiTag {
      const target = stmts.fetchById.get(targetId) as WikiTagRow | undefined;
      if (!target) throw new Error(`Target WikiTag not found: ${targetId}`);

      const doMerge = db.transaction(() => {
        // Move assignments from source to target (ignore duplicates)
        stmts.updateAssignmentTagId.run({
          new_tag_id: targetId,
          old_tag_id: sourceId,
        });
        // Clean up any remaining source assignments (duplicates that couldn't be moved)
        stmts.deleteAssignmentsByTag.run(sourceId);
        // Delete the source tag
        stmts.delete.run(sourceId);
      });
      doMerge();

      return rowToTag(stmts.fetchById.get(targetId) as WikiTagRow);
    },

    fetchTagsForEntity(entityId: string): WikiTag[] {
      return (stmts.fetchTagsForEntity.all(entityId) as WikiTagRow[]).map(
        rowToTag,
      );
    },

    setTagsForEntity(
      entityId: string,
      entityType: string,
      tagIds: string[],
    ): void {
      const setTags = db.transaction(() => {
        stmts.deleteAssignmentsForEntity.run(entityId);
        const now = new Date().toISOString();
        for (const tagId of tagIds) {
          stmts.insertAssignment.run({
            tag_id: tagId,
            entity_id: entityId,
            entity_type: entityType,
            source: "manual",
            created_at: now,
          });
        }
      });
      setTags();
    },

    syncInlineTags(
      entityId: string,
      entityType: string,
      tagNames: string[],
    ): void {
      const sync = db.transaction(() => {
        // Remove old inline assignments for this entity
        stmts.deleteInlineAssignmentsForEntity.run(entityId);

        const now = new Date().toISOString();
        for (const name of tagNames) {
          // Find or create the tag
          let tag = stmts.fetchByName.get(name) as WikiTagRow | undefined;
          if (!tag) {
            const id = `tag-${crypto.randomUUID()}`;
            stmts.insert.run({
              id,
              name,
              color: "#808080",
              created_at: now,
              updated_at: now,
            });
            tag = stmts.fetchById.get(id) as WikiTagRow;
          }
          // Create inline assignment
          stmts.insertAssignment.run({
            tag_id: tag.id,
            entity_id: entityId,
            entity_type: entityType,
            source: "inline",
            created_at: now,
          });
        }
      });
      sync();
    },

    fetchAllAssignments(): WikiTagAssignment[] {
      return (stmts.fetchAllAssignments.all() as WikiTagAssignmentRow[]).map(
        rowToAssignment,
      );
    },

    restoreAssignment(
      tagId: string,
      entityId: string,
      entityType: string,
      source: string,
    ): void {
      const now = new Date().toISOString();
      stmts.insertAssignment.run({
        tag_id: tagId,
        entity_id: entityId,
        entity_type: entityType,
        source,
        created_at: now,
      });
    },
  };
}

export type WikiTagRepository = ReturnType<typeof createWikiTagRepository>;

File: src-tauri/src/db/wiki_tag_repository.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::helpers;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WikiTag {
    pub id: String,
    pub name: String,
    pub color: String,
    pub text_color: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WikiTagAssignment {
    pub tag_id: String,
    pub entity_id: String,
    pub entity_type: String,
    pub source: String,
    pub created_at: String,
}

fn row_to_wiki_tag(row: &rusqlite::Row) -> rusqlite::Result<WikiTag> {
    Ok(WikiTag {
        id: row.get("id")?,
        name: row.get("name")?,
        color: row.get("color")?,
        text_color: row.get("text_color")?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

fn row_to_assignment(row: &rusqlite::Row) -> rusqlite::Result<WikiTagAssignment> {
    Ok(WikiTagAssignment {
        tag_id: row.get("tag_id")?,
        entity_id: row.get("entity_id")?,
        entity_type: row.get("entity_type")?,
        source: row.get("source")?,
        created_at: row.get("created_at")?,
    })
}

pub fn fetch_all(conn: &Connection) -> rusqlite::Result<Vec<WikiTag>> {
    let mut stmt = conn.prepare("SELECT * FROM wiki_tags ORDER BY name")?;
    let rows = stmt.query_map([], |row| row_to_wiki_tag(row))?;
    rows.collect()
}

pub fn search(conn: &Connection, query: &str) -> rusqlite::Result<Vec<WikiTag>> {
    let pattern = format!("%{}%", query);
    let mut stmt =
        conn.prepare("SELECT * FROM wiki_tags WHERE name LIKE ?1 ORDER BY name")?;
    let rows = stmt.query_map([&pattern], |row| row_to_wiki_tag(row))?;
    rows.collect()
}

pub fn create(conn: &Connection, name: &str, color: &str) -> rusqlite::Result<WikiTag> {
    let id = format!("tag-{}", helpers::new_uuid());
    create_with_id(conn, &id, name, color)
}

pub fn create_with_id(
    conn: &Connection,
    id: &str,
    name: &str,
    color: &str,
) -> rusqlite::Result<WikiTag> {
    let now = helpers::now();
    conn.execute(
        "INSERT OR REPLACE INTO wiki_tags (id, name, color, created_at, updated_at) \
         VALUES (?1, ?2, ?3, ?4, ?5)",
        params![id, name, color, &now, &now],
    )?;

    let mut stmt = conn.prepare("SELECT * FROM wiki_tags WHERE id = ?1")?;
    stmt.query_row([id], |row| row_to_wiki_tag(row))
}

pub fn update(conn: &Connection, id: &str, updates: &Value) -> rusqlite::Result<WikiTag> {
    let mut sets = Vec::new();
    let mut values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

    if let Some(v) = updates.get("name") {
        sets.push("name = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("color") {
        sets.push("color = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("textColor") {
        sets.push("text_color = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }

    if sets.is_empty() {
        let mut stmt = conn.prepare("SELECT * FROM wiki_tags WHERE id = ?1")?;
        return stmt.query_row([id], |row| row_to_wiki_tag(row));
    }

    sets.push("updated_at = datetime('now')");
    values.push(Box::new(id.to_string()));

    let sql = format!("UPDATE wiki_tags SET {} WHERE id = ?", sets.join(", "));
    let params: Vec<&dyn rusqlite::types::ToSql> = values.iter().map(|v| v.as_ref()).collect();
    conn.execute(&sql, params.as_slice())?;

    let mut stmt = conn.prepare("SELECT * FROM wiki_tags WHERE id = ?1")?;
    stmt.query_row([id], |row| row_to_wiki_tag(row))
}

pub fn delete(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM wiki_tag_assignments WHERE tag_id = ?1", [id])?;
    conn.execute("DELETE FROM wiki_tags WHERE id = ?1", [id])?;
    Ok(())
}

pub fn merge(
    conn: &Connection,
    source_id: &str,
    target_id: &str,
) -> rusqlite::Result<WikiTag> {
    let tx = conn.unchecked_transaction()?;

    // Move assignments from source to target (skip duplicates)
    tx.execute(
        "INSERT OR IGNORE INTO wiki_tag_assignments (tag_id, entity_id, entity_type, source, created_at) \
         SELECT ?2, entity_id, entity_type, source, created_at \
         FROM wiki_tag_assignments WHERE tag_id = ?1",
        params![source_id, target_id],
    )?;

    // Delete source assignments
    tx.execute(
        "DELETE FROM wiki_tag_assignments WHERE tag_id = ?1",
        [source_id],
    )?;

    // Delete source tag
    tx.execute("DELETE FROM wiki_tags WHERE id = ?1", [source_id])?;

    tx.commit()?;

    let mut stmt = conn.prepare("SELECT * FROM wiki_tags WHERE id = ?1")?;
    stmt.query_row([target_id], |row| row_to_wiki_tag(row))
}

pub fn fetch_tags_for_entity(
    conn: &Connection,
    entity_id: &str,
) -> rusqlite::Result<Vec<WikiTag>> {
    let mut stmt = conn.prepare(
        "SELECT t.* FROM wiki_tags t \
         INNER JOIN wiki_tag_assignments a ON t.id = a.tag_id \
         WHERE a.entity_id = ?1 \
         ORDER BY t.name",
    )?;
    let rows = stmt.query_map([entity_id], |row| row_to_wiki_tag(row))?;
    rows.collect()
}

pub fn set_tags_for_entity(
    conn: &Connection,
    entity_id: &str,
    entity_type: &str,
    tag_ids: &[String],
) -> rusqlite::Result<()> {
    let tx = conn.unchecked_transaction()?;

    tx.execute(
        "DELETE FROM wiki_tag_assignments WHERE entity_id = ?1 AND source = 'manual'",
        [entity_id],
    )?;

    let now = helpers::now();
    for tag_id in tag_ids {
        tx.execute(
            "INSERT OR IGNORE INTO wiki_tag_assignments \
             (tag_id, entity_id, entity_type, source, created_at) \
             VALUES (?1, ?2, ?3, 'manual', ?4)",
            params![tag_id, entity_id, entity_type, &now],
        )?;
    }

    tx.commit()
}

pub fn sync_inline_tags(
    conn: &Connection,
    entity_id: &str,
    entity_type: &str,
    tag_names: &[String],
) -> rusqlite::Result<()> {
    let tx = conn.unchecked_transaction()?;
    let now = helpers::now();

    // Find or create each tag
    let mut tag_ids = Vec::new();
    for name in tag_names {
        let existing: Option<String> = tx
            .query_row(
                "SELECT id FROM wiki_tags WHERE name = ?1",
                [name],
                |row| row.get(0),
            )
            .ok();

        let tag_id = match existing {
            Some(id) => id,
            None => {
                let id = format!("tag-{}", helpers::new_uuid());
                tx.execute(
                    "INSERT INTO wiki_tags (id, name, color, created_at, updated_at) \
                     VALUES (?1, ?2, '#6B7280', ?3, ?4)",
                    params![&id, name, &now, &now],
                )?;
                id
            }
        };
        tag_ids.push(tag_id);
    }

    // Remove old inline assignments
    tx.execute(
        "DELETE FROM wiki_tag_assignments WHERE entity_id = ?1 AND source = 'inline'",
        [entity_id],
    )?;

    // Insert new inline assignments
    for tag_id in &tag_ids {
        tx.execute(
            "INSERT OR IGNORE INTO wiki_tag_assignments \
             (tag_id, entity_id, entity_type, source, created_at) \
             VALUES (?1, ?2, ?3, 'inline', ?4)",
            params![tag_id, entity_id, entity_type, &now],
        )?;
    }

    tx.commit()
}

pub fn fetch_all_assignments(conn: &Connection) -> rusqlite::Result<Vec<WikiTagAssignment>> {
    let mut stmt = conn.prepare(
        "SELECT * FROM wiki_tag_assignments ORDER BY created_at",
    )?;
    let rows = stmt.query_map([], |row| row_to_assignment(row))?;
    rows.collect()
}

pub fn restore_assignment(
    conn: &Connection,
    tag_id: &str,
    entity_id: &str,
    entity_type: &str,
    source: &str,
) -> rusqlite::Result<()> {
    let now = helpers::now();
    conn.execute(
        "INSERT OR IGNORE INTO wiki_tag_assignments \
         (tag_id, entity_id, entity_type, source, created_at) \
         VALUES (?1, ?2, ?3, ?4, ?5)",
        params![tag_id, entity_id, entity_type, source, &now],
    )?;
    Ok(())
}
